### ISSUE TITLE
oracle: PL/SQL parse failure

### ISSUE THREAD

**Issue by deliciouslytyped** [tobymao/sqlglot#1356]
I get the impression from other issues that the various procedural dialects are also intended to be supported.
The following homework related code fails to parse with the Oracle dialect:

```
declare 
y varchar2(4); y2 varchar2(4);
begin
y:=to_char(sysdate,'yyyy');
if to_char(sysdate,'mm')<='08'
  then y:=y-1;
end if;
select to_char(max(megrendeles_datuma),'yyyy') into y2 from s_megrendeles;
y:=y-y2;

update s_megrendeles
set megrendeles_datuma=add_months(megrendeles_datuma, 12*y);

update s_ugyfel
set szul_dat=add_months(szul_dat, 12*y);

update s_ut
set indulasi_ido=add_months(indulasi_ido, 12*y),
erkezesi_ido=add_months(erkezesi_ido, 12*y);
commit;
end;
```

```
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.2032.0_x64__qbz5n2kfra8p0\lib\code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "C:\Users\MyUser\PycharmProjects\flaskProject1\venv\lib\site-packages\sqlglot\dialects\dialect.py", line 163, in parse
    return self.parser(**opts).parse(self.tokenize(sql), sql)
  File "C:\Users\MyUser\PycharmProjects\flaskProject1\venv\lib\site-packages\sqlglot\parser.py", line 799, in parse
    return self._parse(
  File "C:\Users\MyUser\PycharmProjects\flaskProject1\venv\lib\site-packages\sqlglot\parser.py", line 865, in _parse
    self.raise_error("Invalid expression / Unexpected token")
  File "C:\Users\MyUser\PycharmProjects\flaskProject1\venv\lib\site-packages\sqlglot\parser.py", line 908, in raise_error
    raise error
sqlglot.errors.ParseError: Invalid expression / Unexpected token. Line 2, Col: 3.
  declare 
y varchar2(4); y2 varchar2(4);
begin
y:=to_char(sysdate,'yyyy');
if to_char(sysdate,'mm')<='08'
  then y:=y-1;```

**Comment by GeorgeSittas** [comment]
> I get the impression from other issues that the various procedural dialects are also intended to be supported.

This isn't in SQLGlot's scope, as it mostly aims to cover standard SQL (or at least "SQL-ish" dialects that maintain the declarative nature of the language). Just to make sure we're aligned on this, not closing this as not planned yet.

cc: @tobymao

**Comment by deliciouslytyped** [comment]
Ah, Ok :+1:

**Comment by deliciouslytyped** [comment]
It would be nice if there was some table of intended and supported language levels and features, though I imagine part of the reason this doesn't already exist is that being explicit about this hasn't been a priority?

**Comment by tobymao** [comment]
yes agreed. in general sqlglot is used to power analytical use cases.

**Comment by deliciouslytyped** [comment]
Sorry for the slight side-track - I want to use it for rewriting SQL queries. Is this in general the best library for the job?

More specifically I'm looking to use it for non-hardened sandboxing in an educational SQL learning environment / problem set environment, mostly by rewriting table names to/from a prefixed form. This is to multiplex the tables in the problem sets without needing to create users for isolation, because I don't have permissions to do that in the database. (This would be the "normal" solution.)

This is the context in which the issues I'm submitting come up. I'm trying to port the SQL code I'm running into to my "framework".

**Comment by tobymao** [comment]
this is probably the best library i'm aware of for query rewriting

**Comment by deliciouslytyped** [comment]
I forgot to mention that a key point is this means I'm "round-tripping" the Oracle SQL syntax through the library; parsing it, editing the AST, and then .sql()-ing it. This is the appropriate way to do this? Can I expect this to generally work?

Thanks again!

**Comment by tobymao** [comment]
yes, the caveat being oracle doesn't have top tier support. i've never used it before and i have no way to test it.

**Comment by deliciouslytyped** [comment]
Could the procedural portions of dialects be supported to the extent that it doesn't break parsing if there is procedural code in a file?

**Comment by tobymao** [comment]
you can probably handle that yourself using regex, or a custom sqlglot declare -> command tokenizer

**Comment by jarodsmk** [comment]
Thought I'd add to this - I ended up writing code to try & parse my Oracle PL/SQL scripts without knowing that it wasn't supported, only to have it fail when trying to parse the DECLARE block variable types. 

My use case is I'd like to run through a variety of files and identify table & column names that exceed a certain character limit, something that changes around the 12c version. 

Prior to this I had a simple NodeJS regex script implemented, but to get around certain caveats like strings within UPDATE/INSERT statements proved to be a pain, hence why I looked for an AST/parser solution leading me to sqlglot.

I think a blurb in the readme describing PL code as being unsupported would be valuable :)
